cdap-master/src/main/java/co/cask/cdap/data/tools/LogUpgrader.java

```diff
@@ -0,0 +1,203 @@
+                throw Throwables.propagate(t);
+              }
+            } else if (!isKeyValid(key, OTHERS)) {
+              LOG.warn("Unknown key {} found. Failed to Upgrade", key);
+            }
+          }
+        }
+        return null;
+      }
+    });
+  }
+
+  /**
+   * Adds the namespace in the key which is the log partition
+   *
+   * @param key the key which is the log partition
+   * @return the new key with namespace
+   */
+  private String upgradeKey(String key) {
+    if (key.startsWith(Constants.Logging.SYSTEM_NAME)) {
+      return key.replace(Constants.Logging.SYSTEM_NAME, Constants.SYSTEM_NAMESPACE);
+    }
+    return key.replace(DEVELOPER_STRING, Constants.DEFAULT_NAMESPACE);
+  }
+
+  /**
+   * Creates a new path depending on the old log file path
+   *
+   * @param key         the key for this log meta entry
+   * @param oldLocation the old log {@link Location}
+   * @return the {@link Location}
+   * @throws IOException
+   * @throws URISyntaxException
+   */
+  private Location upgradePath(String key, String oldLocation) throws IOException, URISyntaxException {
+    Location location = locationFactory.create(new URI(oldLocation));
+    Location newLocation = null;
+    if (key.startsWith(Constants.Logging.SYSTEM_NAME)) {
+      if (location != null) {
+        // Example path of this type: hdfs://cdap28-161119-1000.dev.continuuity.
+        // net/cdap/logs/avro/cdap/services/service-appfabric/2015-02-26/1424988452088.avro
+        // removes the second occurrence of "cdap/" in the path
+        newLocation = updateLogLocation(location, Constants.SYSTEM_NAMESPACE);
+        // newLocation will be: hdfs://cdap28-161119-1000.dev.continuuity.net/cdap/system/logs/avro/
+        // services/service-appfabric/2015-02-26/1424988452088.avro
+      }
+    } else if (key.startsWith(DEVELOPER_STRING)) {
+      if (location != null) {
+        // Exmaple of this type: hdfs://cdap28-161119-1000.dev.continuuity.net/cdap/logs/avro/developer/HelloWorld/
+        // flow-WhoFlow/2015-02-26/1424988484082.avro
+        newLocation = updateLogLocation(location, Constants.DEFAULT_NAMESPACE);
+        // newLocation will be: dfs://cdap28-161119-1000.dev.continuuity.net/cdap/default/logs/avro/HelloWorld/
+        // flow-WhoFlow/2015-02-26/1424988484082.avro
+      }
+    }
+    return newLocation;
+  }
+
+  /**
+   * Strips different parts from the old log location and creates a new one
+   *
+   * @param location  the old log {@link Location}
+   * @param nameSpace the namespace which will be added to the new log location
+   * @return the log {@link Location}
+   * @throws IOException
+   */
+  private Location updateLogLocation(Location location, String nameSpace) throws IOException {
+    String logFilename = location.getName();
+    Location parentLocation = Locations.getParent(location);  // strip filename
+    String date = parentLocation.getName();
+    parentLocation = Locations.getParent(parentLocation); // strip date
```

Comment:
Each of these ``Locations.getParent(parentLocation)`` may return a null, theoretically. However, it should never happen since this is the expected path from the log.meta table. May be worth precondition'ing them.